source/opt/decoration_manager.cpp

```diff
@@ -46,6 +46,21 @@ void DecorationManager::RemoveDecorationsFrom(
   for (ir::Instruction* inst : decorations_info.indirect_decorations) {
     assert(inst->opcode() == SpvOpGroupDecorate ||
            inst->opcode() == SpvOpGroupMemberDecorate);
+
+    std::vector<ir::Instruction*> group_decorations_to_keep;
+    const uint32_t group_id = inst->GetSingleWordInOperand(0u);
+    const auto group_iter = id_to_decoration_insts_.find(group_id);
+    for (ir::Instruction* group_decoration :
+         group_iter->second.direct_decorations)
+      if (!pred(*group_decoration))
+        group_decorations_to_keep.push_back(group_decoration);
+
+    // If all decorations should be kept, move to the next group
+    if (group_decorations_to_keep.size() ==
+        group_iter->second.direct_decorations.size())
+      continue;
```

Comment:
If you are keeping all of the decorations, should you clear `group_decorations_to_keep` so you don't add them below?

Reply:
You are correct, I need to do that. I’ll add a test that would fail with the current code, to detect possible regressions and check that the code is properly fixed.

Reply:
Actually, looking at the code again, they won’t be added later on (I assume you are thinking of the code [right here](https://github.com/KhronosGroup/SPIRV-Tools/pull/1373/files/91f73ab072b9176e4e539a343d868f005847e8a4#diff-b567013ae7fa0970645039363ebc5cd7R99)): `group_decorations_to_keep` is local to the loop, so when continuing and going to the next loop iteration, `group_decorations_to_keep` will be a new, empty, vector.

I added two new tests to check that behaviour, which uncovered a bug where the medata was not updated with the clone instructions due to calling `AnalyzeDefUse()` (which does not update decorations) instead of `AnalyzeUses()`.
I alos added a test for the `CloneDecorations()` which was suffering from the same issue, as well as from an other bug (the list of direct decorations from groups was modified as we were iterating over it).